test: refactor Requests retry helper and make it generic

Ebaneck opened the pull request:
**Component**:

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->
'test'

**Context**: 

See: https://github.com/scality/metalk8s/pull/2249#discussion_r381489467

**Summary**:

Made the requests retry helper generic by declaring it in `tests/utils.py` before importing it into the authentication test.

**Acceptance criteria**: 


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #ISSUE_NUMBER

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

Ebaneck committed:
test: refactor retry helper and make it generic

Initially, we declare the requests retry helper within the authentication
tests but this helper could be made generic such that other tests could
utilize it.

Now, the request retry helper is defined in `test/utils.py` before being
imported into the authentication tests.

bert-e commented:
# Hello ebaneck,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/fix-code-style-comments` into
`development/2.5`
* [w/2.6/improvement/fix-code-style-comments](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/fix-code-style-comments) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

gdemonet commented on tests/utils.py:
```diff
@@ -61,3 +65,34 @@ def get_node_name(nodename, ssh_config=None):
+    This helper performs 3 retries with an exponential sleep interval between
+    each request.
+
+    We only retry internal server errors(500) & service unavailable errors(503)
```
Those are default values, given this function, it seems we can override many arguments. Docstring should explain this, and maybe explain the argument values?

slaperche-scality commented on tests/utils.py:
```diff
@@ -61,3 +65,34 @@ def get_node_name(nodename, ssh_config=None):
+
+
+# Source: https://www.peterbe.com/plog/best-practice-with-retries-with-requests
+
```
you can remove this empty line

slaperche-scality commented on tests/utils.py:
```diff
@@ -61,3 +65,34 @@ def get_node_name(nodename, ssh_config=None):
+    each request.
+
+    We only retry internal server errors(500) & service unavailable errors(503)
+
```
you can remove this empty line

gdemonet replied:
Still here...

slaperche-scality commented on tests/utils.py:
```diff
@@ -61,3 +65,34 @@ def get_node_name(nodename, ssh_config=None):
+    This helper performs 3 retries with an exponential sleep interval between
+    each request.
+
+    We only retry internal server errors(500) & service unavailable errors(503)
```
This is the default behavior (so you could start the sentence with "By default this helper…"

Also would be nice to have doc for the arguments, something like that:

```python
    """Configure a `requests.Session` for retry on error.

    By default this helper performs 3 retries with an exponential sleep interval
    between each request and only retry internal server errors(500) & service
    unavailable errors(503).

    Arguments:
        retries:          […]
        backoff_factor:   […]
        status_forcelist: […]
        method_whitelist: […]
        session:          […]

    Returns:
        A `requests.Session` object configured for retry.
    """
```

Ebaneck replied:
Done

gdemonet reviewed (changes requested):
Small change on the docstring to make it more obvious that this helper is configurable.

If you really want to make it super easy to use, you could make it a Pytest fixture:

```python
@pytest.fixture(scope="function")
def request_retry_session(request):
    # Callers can inject arguments using `pytest.mark.parametrize`
    params = getattr(request, 'param', {})

    return utils.request_retry_session(**params)


def test_flaky_request(request_retry_session):
    resp = request_retry_session.get("/flaky")
    assert resp == 'OK'


@pytest.mark.parametrize('request_retry_session', [
    {'retries': 1, 'backoff_factor': 0.5},
], indirect=True)
def test_stable_request(request_retry_session):
    resp = request_retry_session.get("/stable")
    assert resp == 'OK'
```

Although I think I remember using `pytest.mark.parametrize` for fixtures in `pytest-bdd` scenarii is kind of painful...

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

bert-e commented:
# History mismatch



Merge commit #ff456f03316acda53f2e849547fa1442328fe3db on the integration branch
`w/2.6/improvement/fix-code-style-comments` is merging a branch which is neither the current
branch `improvement/fix-code-style-comments` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `improvement/fix-code-style-comments` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Ebaneck committed:
test: make request retry helper a pytest fixture

This commit adds `request retry helper` as a pytest fixture
ensuring that it is easy to re-use in whatsoever test function

bert-e commented:
# History mismatch



Merge commit #5b16e9806f634aa9b921e14cfbff5bd33c93db8e on the integration branch
`w/2.6/improvement/fix-code-style-comments` is merging a branch which is neither the current
branch `improvement/fix-code-style-comments` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `improvement/fix-code-style-comments` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/fix-code-style-comments` into
`development/2.5`
* [w/2.6/improvement/fix-code-style-comments](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/fix-code-style-comments) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

bert-e commented:
# History mismatch



Merge commit #2b61e32bcdb2958b3fca1c20eb0db4afd625f419 on the integration branch
`w/2.6/improvement/fix-code-style-comments` is merging a branch which is neither the current
branch `improvement/fix-code-style-comments` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `improvement/fix-code-style-comments` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Ebaneck commented:
/reset

Ebaneck commented:
/approve

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/fix-code-style-comments` into
`development/2.5`
* [w/2.6/improvement/fix-code-style-comments](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/fix-code-style-comments) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet







*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/aa28bdde781b27f7814820043466da973d3d1fae) did not succeed in branch w/2.6/improvement/fix-code-style-comments.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



Please check the status of the associated issue None.

Goodbye ebaneck.
